modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -1575,33 +1608,38 @@ private CompletableFuture<Void> finishAndCleanup(
+            // - A coordinator might send a commit, then die, but the commit message might still arrive at the commit partition primary.
+            //   If it arrived with a delay, another node might come across a write intent/lock from that tx
+            //   and realize that the coordinator is no longer available and start tx recovery.
+            //   The original commit message might arrive later than the recovery one,
+            //   hence a 'commit over rollback' case.
+            // The possible states that a 'commit' is allowed to see:
             // - null (if it's the first change state attempt)
             // - committed (if it was already updated in the previous attempt)
             // - aborted (if it was aborted by the initiate recovery logic,
             //   though this is a very unlikely case because initiate recovery will only roll back the tx if coordinator is dead).
             //
             // Within 'roll back' it's allowed to see:
-            // - null
-            // - aborted
-            // Other combinations of states are not possible.
-
-            // First, throw an exception if we are trying to abort an already committed tx.
-            assert !(txMeta.txState() == COMMITTED && !commit) : "Not allowed to abort an already committed transaction.";
+            // - null (if it's the first change state attempt)
+            // - aborted  (if it was already updated in the previous attempt or the result of a concurrent recovery)
+            // - commit (if initiate recovery has started, but a delayed message from the coordinator finally arrived and executed earlier).
 
-            // If a 'commit' sees a tx in the ABORTED state (valid as per the explanation above), let the client know with an exception.
-            if (commit && txMeta.txState() == ABORTED) {
-                LOG.error("Failed to commit a transaction that is already aborted [txId={}].", txId);
+            // Let the client know a transaction has finished with a different outcome.
+            if (commit != (txMeta.txState() == COMMITTED)) {
+                LOG.error("Failed to finish a transaction that is already finished [txId={}, expectedState={}, actualState={}].",
+                        txId,
+                        commit ? COMMITTED : ABORTED,
+                        txMeta.txState()
+                );
 
-                throw new TransactionException(TX_WAS_ABORTED_ERR,
-                        "Failed to change the outcome of a finished transaction"
-                                + " [txId=" + txId + ", txState=" + txMeta.txState() + "].");
+                throw new TransactionAlreadyFinishedException(
+                        "Failed to change the outcome of a finished transaction [txId=" + txId + "].",
+                        new TransactionResult(txMeta.txState(), txMeta.commitTimestamp())
+                );
```

Comment:
Don't you think that the exception message should include the transaction result?